src/sum_products.rs

```diff
@@ -1,28 +1,68 @@
-use crate::npyffi::{NPY_CASTING, NPY_ORDER};
-use crate::{Element, PyArray, PY_ARRAY_API};
+use std::borrow::Cow;
+use std::ffi::{CStr, CString};
+use std::ptr::null_mut;
+
 use ndarray::{Dimension, IxDyn};
-use pyo3::{AsPyPointer, FromPyPointer, PyAny, PyNativeType, PyResult};
-use std::ffi::CStr;
+use pyo3::{AsPyPointer, FromPyObject, FromPyPointer, PyAny, PyNativeType, PyResult};
+
+use crate::array::PyArray;
+use crate::dtype::Element;
+use crate::npyffi::{array::PY_ARRAY_API, NPY_CASTING, NPY_ORDER};
+
+/// Return value of a function that can yield either an array or a scalar.
+pub trait ArrayOrScalar<'py, T>: FromPyObject<'py> {}
```

Comment:
I chose to go for a trait in return position so that the current scheme of selecting the result type via type inference continues to work. (Compared to say using an `enum`.)

We might want to consider sealing the trait, but since the trait itself is mainly documentation (I could have use a `FromPyObject` bound directly on the methods) and nothing bad will happen if anyone implements this for other types, I decided against the additionally code noisy necessary for sealing.